### DISCUSSION TITLE
Download Manager

### DISCUSSION THREAD

**Discussion by Nocccer** [Heroic-Games-Launcher/HeroicGamesLauncher#1735] [Ideas]
# UML:
![image](https://user-images.githubusercontent.com/61798668/184882968-8fedc6b5-148d-4827-a4f7-941122af9282.png)

# Squence:
![image](https://user-images.githubusercontent.com/61798668/184887221-94c83e1e-518c-44cd-aa0a-83a6006f7b22.png)
Draw.io file: 
[Download-Manager-Sequence.drawio.zip](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/files/9351258/Download-Manager-Sequence.drawio.zip)

**Comment by arielj** [comment]
I would change a bit the flow (sorry it's just text xD):

At Boot / When an installation finished / When an installation is cancelled:
- check queue
- if empty, do nothing
- if not empty, start installing the first element

When an installation is requested:
- check if already installing something else
- if not installing, enqueue and start installing
- if installing, just enqueue

That way we don't need a loop checking if the queue (at least that's what I understand from the `Queue empty?` arrows, that it's a loop).

I think we should also store the status (waiting / paused / installing / installed / canceled) so we can show a log.

**Reply by Nocccer** [reply]
Actually my sequenz diagram is exactly that. The uml above is more generic.

**Reply by arielj** [reply]
oh ok! I tought the first one was a loop checking the empty queue because of the arrows comming back! nevermind then :+1:

**Reply by Nocccer** [reply]
The first one is really bad actually. Second one explains a little bit more what happens during runtime. It does not show all behaviors like cancel or resume.
